airflow/models/dag_version.py

```diff
@@ -0,0 +1,142 @@
+
+    from airflow.models.dagcode import DagCode
+    from airflow.models.serialized_dag import SerializedDagModel
+
+log = logging.getLogger(__name__)
+
+
+class DagVersion(Base):
+    """Model to track the versions of DAGs in the database."""
+
+    __tablename__ = "dag_version"
+    id = Column(Integer, primary_key=True)
+    version_number = Column(Integer)
+    version_name = Column(StringID())
+    dag_id = Column(StringID(), ForeignKey("dag.dag_id", ondelete="CASCADE"))
+    dag_model = relationship("DagModel", back_populates="dag_versions")
+    dag_code = relationship("DagCode", back_populates="dag_version", uselist=False)
+    serialized_dag = relationship("SerializedDagModel", back_populates="dag_version", uselist=False)
+    dag_runs = relationship("DagRun", back_populates="dag_version")
+    task_instances = relationship("TaskInstance", back_populates="dag_version")
+
+    def __init__(
+        self,
+        *,
+        dag_id: str,
+        version_number: int,
+        dag_code: DagCode,
+        serialized_dag: SerializedDagModel,
+        version_name: str | None = None,
+    ):
+        self.dag_id = dag_id
+        self.version_number = version_number
+        self.dag_code = dag_code
+        self.serialized_dag = serialized_dag
+        self.version_name = version_name
+
+    def __repr__(self):
+        return f"<DagVersion {self.dag_id} - {self.version_name}>"
+
+    @classmethod
+    def _generate_random_string(cls):
+        letters = string.ascii_letters + string.digits
+        return "dag-" + "".join(random.choice(letters) for i in range(10))
+
+    @classmethod
+    @provide_session
+    def _generate_unique_random_string(cls, session: Session = NEW_SESSION):
+        while True:
+            random_str = cls._generate_random_string()
+            # Check if the generated string exists
+            if not session.scalar(select(cls).where(cls.version_name == random_str)):
+                return random_str
+
+    @classmethod
+    @provide_session
+    def write_dag_version(
+        cls,
+        *,
+        dag_id: str,
+        dag_code: DagCode,
+        serialized_dag: SerializedDagModel,
+        version_name: str | None = None,
+        session: Session = NEW_SESSION,
+    ):
+        """Write a new DagVersion into database."""
+        existing_dag_version = session.scalar(
+            select(cls).where(cls.dag_id == dag_id).order_by(cls.version_number.desc()).limit(1)
+        )
+        version_number = 1
+
+        if existing_dag_version:
+            version_number = existing_dag_version.version_number + 1
+        if not version_name and existing_dag_version:
+            version_name = existing_dag_version.version_name
+
+        dag_version = DagVersion(
+            dag_id=dag_id,
+            version_number=version_number,
+            dag_code=dag_code,
+            serialized_dag=serialized_dag,
+            version_name=version_name or cls._generate_unique_random_string(session),
```

Comment:
I think we should augment the `version_name` we store when a DAG author provides is to us. Otherwise, we'd have the same `version_name` for all versions until they change it, which isn't great.

e.g. if I have `version_name=foo` in my DAG, and I add a new task but don't update the version. So maybe something like `foo-{version_number}` or something? Or maybe we toss on an int to signify when it hasn't changed, but the DAG has, like `foo-2`.

We also only stamp a new version when the serdag actually changes. I wonder if we'd be better off combing the user provided version and the serdag hash to determine when to create a new version?

Reply:
> I think we should augment the version_name we store when a DAG author provides is to us. Otherwise, we'd have the same version_name for all versions until they change it, which isn't great.

>e.g. if I have version_name=foo in my DAG, and I add a new task but don't update the version. So maybe something like foo-{version_number} or something? Or maybe we toss on an int to signify when it hasn't changed, but the DAG has, like foo-2.

That's my plan, but I want to have it in a method. The `version_number` increments, while the version_name remains the same until changed. However, a method named `version` would return `version_name-version_number` (`foo-2`), which would be used on the UI. I think having them separate is much better than using Python logic, which might go wrong. 

>We also only stamp a new version when the serdag actually changes. I wonder if we'd be better off combing the user provided version and the serdag hash to determine when to create a new version?

The serdag is 1:1 to dag_version. When it changes, the version also changes and there's a link between the two tables which allows us to access the hash e.g :
```python
version = session.query(DagVersion).get(1)
print(version.serialized_dag.dag_hash)
```
I feel we shouldn't combine the hash and version since a change in the user-provided version name would trigger a new serialized dag version, which also means a new dagversion.

The dag_hash can remain an internal property that we use to know that the version has changed